relays/utils/src/relay_loop.rs

```diff
@@ -32,60 +34,164 @@ pub trait Client: Clone + Send + Sync {
 	async fn reconnect(&mut self) -> Result<(), Self::Error>;
 }
 
-/// Run relay loop.
-///
-/// This function represents an outer loop, which in turn calls provided `loop_run` function to do
-/// actual job. When `loop_run` returns, this outer loop reconnects to failed client (source,
-/// target or both) and calls `loop_run` again.
-pub async fn run<SC: Client, TC: Client, R, F>(
-	reconnect_delay: Duration,
-	mut source_client: SC,
-	mut target_client: TC,
-	loop_run: R,
-) where
-	R: Fn(SC, TC) -> F,
-	F: Future<Output = Result<(), FailedClient>>,
-{
-	loop {
-		let result = loop_run(source_client.clone(), target_client.clone()).await;
-
-		match result {
-			Ok(()) => break,
-			Err(failed_client) => loop {
-				async_std::task::sleep(reconnect_delay).await;
-				if failed_client == FailedClient::Both || failed_client == FailedClient::Source {
-					match source_client.reconnect().await {
-						Ok(()) => (),
-						Err(error) => {
-							log::warn!(
-								target: "bridge",
-								"Failed to reconnect to source client. Going to retry in {}s: {:?}",
-								reconnect_delay.as_secs(),
-								error,
-							);
-							continue;
+/// Returns generic loop that may be customized and started.
+pub fn relay_loop<SC, TC>(source_client: SC, target_client: TC) -> Loop<SC, TC, ()> {
+	Loop {
+		source_client,
+		target_client,
+		loop_metric: None,
+	}
+}
+
+/// Generic relay loop.
+pub struct Loop<SC, TC, LM> {
+	source_client: SC,
+	target_client: TC,
+	loop_metric: Option<LM>,
+}
+
+/// Relay loop metrics builder.
+pub struct LoopMetrics<SC, TC, LM> {
+	relay_loop: Loop<SC, TC, ()>,
+	registry: Registry,
+	loop_metric: Option<LM>,
+}
+
+impl<SC, TC, LM> Loop<SC, TC, LM> {
+	/// Start building loop metrics using given prefix.
+	///
+	/// Panics if `prefix` is empty.
+	pub fn with_metrics(self, prefix: String) -> LoopMetrics<SC, TC, ()> {
+		assert!(!prefix.is_empty(), "Metrics prefix can not be empty");
+
+		LoopMetrics {
+			relay_loop: Loop {
+				source_client: self.source_client,
+				target_client: self.target_client,
+				loop_metric: None,
+			},
+			registry: Registry::new_custom(Some(prefix), None)
+				.expect("only fails if prefix is empty; prefix is not empty; qed"),
+			loop_metric: None,
+		}
+	}
+
+	/// Run relay loop.
+	///
+	/// This function represents an outer loop, which in turn calls provided `loop_run` function to do
+	/// actual job. When `loop_run` returns, this outer loop reconnects to failed client (source,
+	/// target or both) and calls `loop_run` again.
+	pub async fn run<R, F>(mut self, loop_run: R) -> Result<(), String>
```

Comment:
```suggestion
	pub async fn run<R, F>(mut self, run_loop: R) -> Result<(), String>
```

Reply:
From what I understand this doesn't need to be any sort of loop though, it really just needs to be something which returns a `Future`, right? Maybe we shouldn't even have loop in the name here

Reply:
Yes, you may use it to run arbitrary futures, I see no point in using abstract names here. It is only used (and most probably will only be used) to run relay loops. I bet you'll be the first one to object if I'll rename this whole functionality to something like `fn run_future_with_retry()` :)